go.mod

```diff
@@ -1,60 +1,82 @@
+	k8s.io/apimachinery v0.19.4
+	k8s.io/apiserver v0.19.2
+	k8s.io/autoscaler/cluster-autoscaler v0.0.0-20201102145256-1ba66f5cf714
+	k8s.io/client-go v11.0.1-0.20190409021438-1a26190bd76a+incompatible
+	k8s.io/component-base v0.19.2
+	k8s.io/klog v1.0.0
+	k8s.io/kubernetes v1.19.1
 	sigs.k8s.io/application v0.8.3
 )
 
+// openshift cluster api for cluster-api based node provision and suspend
+require github.com/openshift/machine-api-operator v0.2.1-0.20201216110516-d9e48bb9fc0b
+
+replace (
+	sigs.k8s.io/cluster-api-provider-aws => github.com/openshift/cluster-api-provider-aws v0.2.1-0.20201125052318-b85a18cbf338
+	sigs.k8s.io/cluster-api-provider-azure => github.com/openshift/cluster-api-provider-azure v0.1.0-alpha.3.0.20201130182513-88b90230f2a4
+)
+
+// azure adk to match cluster autoscaler dependency
+replace github.com/Azure/azure-sdk-for-go => github.com/Azure/azure-sdk-for-go v43.0.0+incompatible
+
+// both cluster autoscaler and k8s are not designed to be vendored and use k8s
+// staging dependencies from the core k/k repo. We need to mandatorily provide the
+// replace directives to ensure the right package is pulled wrt k8s and cluster autoscaler
+// cluster autoscaler's hash maps to k8s release 1.19.1 and CAs release "Cluster Autoscaler 1.19.1".
 replace (
-	k8s.io/api => k8s.io/api v0.0.0-20190620084959-7cf5895f2711
-	k8s.io/apiextensions-apiserver => k8s.io/apiextensions-apiserver v0.0.0-20190620085554-14e95df34f1f
-	k8s.io/apimachinery => k8s.io/apimachinery v0.0.0-20190612205821-1799e75a0719
-	k8s.io/apiserver => k8s.io/apiserver v0.0.0-20190620085212-47dc9a115b18
-	k8s.io/cli-runtime => k8s.io/cli-runtime v0.0.0-20190620085706-2090e6d8f84c
-	k8s.io/client-go => k8s.io/client-go v0.0.0-20190620085101-78d2af792bab
-	k8s.io/cloud-provider => k8s.io/cloud-provider v0.0.0-20190620090043-8301c0bda1f0
-	k8s.io/cluster-bootstrap => k8s.io/cluster-bootstrap v0.0.0-20190620090013-c9a0fc045dc1
-	k8s.io/code-generator => k8s.io/code-generator v0.0.0-20190612205613-18da4a14b22b
-	k8s.io/component-base => k8s.io/component-base v0.0.0-20190620085130-185d68e6e6ea
-	k8s.io/cri-api => k8s.io/cri-api v0.0.0-20190531030430-6117653b35f1
-	k8s.io/csi-translation-lib => k8s.io/csi-translation-lib v0.0.0-20190620090116-299a7b270edc
-	k8s.io/kube-aggregator => k8s.io/kube-aggregator v0.0.0-20190620085325-f29e2b4a4f84
-	k8s.io/kube-controller-manager => k8s.io/kube-controller-manager v0.0.0-20190620085942-b7f18460b210
-	k8s.io/kube-proxy => k8s.io/kube-proxy v0.0.0-20190620085809-589f994ddf7f
-	k8s.io/kube-scheduler => k8s.io/kube-scheduler v0.0.0-20190620085912-4acac5405ec6
-	k8s.io/kubelet => k8s.io/kubelet v0.0.0-20190620085838-f1cb295a73c9
-	k8s.io/legacy-cloud-providers => k8s.io/legacy-cloud-providers v0.0.0-20190620090156-2138f2c9de18
-	k8s.io/metrics => k8s.io/metrics v0.0.0-20190620085625-3b22d835f165
-	k8s.io/sample-apiserver => k8s.io/sample-apiserver v0.0.0-20190620085408-1aef9010884e
+	k8s.io/api => k8s.io/api v0.19.1
```

Comment:
I do not quite understand why we need the `replace` directive here? Why can't we put everything in the `require` directive? For example, the `k8s.io/api` is already defined in the `require` directive with version v0.19.4, but is now being replaced with a different version v0.19.1, is it to make it consistent with all other `k8s.io` modules?

Reply:
Talked with @irfanurrehman, this is needed as a workaround to manage dependency for k8s and cluster autoscaler.